src/funnels/Swapper.sol

```diff
@@ -0,0 +1,243 @@
+        else revert("Swapper/file-unrecognized-param");
+        emit File(what, data, val);
+    }
+
+    function setWeights(Weight[] memory newWeights) external toll {
+        uint256 cumPct;
+        uint256[] memory arr = new uint256[](newWeights.length);
+        for(uint256 i; i < newWeights.length;) {
+            (address _box, uint256 _pct) = (newWeights[i].box, uint256(newWeights[i].wad));
+            require(boxes[_box] == 1, "Swapper/invalid-box");
+            require(_pct <= WAD, "Swapper/not-wad-percentage");
+            arr[i] = (uint256(uint160(_box)) << 96) | _pct;
+            cumPct += _pct;
+            unchecked { ++i; }
+        }
+        require(cumPct == WAD, "Swapper/total-weight-not-wad");
+        weights = arr;
+    }
+
+    function getWeightAt(uint256 i) public view returns (address box, uint256 percent) {
+        uint256 weight = weights[i];
+        (box, percent) = (address(uint160(weight >> 96)), weight & (2**96 - 1)); // TODO: check that this tuple assignment results in only one SLOAD
+        require(boxes[box] == 1, "Swapper/invalid-box"); // sanity check that any deauthorised box has also been removed from the weights vector
+    }
+
+    function getWeightsLength() external view returns (uint256 len) {
+        len = weights.length;
+    }
+
+    function swapIn(uint256 wad, uint256 min) external keeper returns (uint256 out) {
+        require(block.timestamp >= zzz + hop, "Swapper/too-soon");
+        zzz = block.timestamp;
+
+        require(wad <= maxIn, "Swapper/wad-too-large");
+        require(min >= wad * wantIn / 10**30 , "Swapper/min-too-small"); // 1/10**30 = 1/WAD * 1/10**12
+
+        BufferLike(buffer).take(address(this), wad);
+
+        bytes memory path = abi.encodePacked(nst, uint24(fee), gem);
+        SwapRouterLike.ExactInputParams memory params = SwapRouterLike.ExactInputParams({
+            path:             path,
+            recipient:        address(this),
+            deadline:         block.timestamp,
+            amountIn:         wad,
+            amountOutMinimum: min
+        });
+        out = SwapRouterLike(uniV3Router).exactInput(params);
+
+        uint256 pending = out;
+        uint256 len = weights.length;
+        for(uint256 i; i < len;) {
+            (address _to, uint256 _percent) = getWeightAt(i);
+            uint256 _amt = (i == len - 1) ? pending : out * _percent / WAD;
+
+            require(GemLike(gem).transfer(_to, _amt));
+            BoxLike(_to).deposit(gem, _amt);
+
+            pending -= _amt;
+            unchecked { ++i; }
+        }
+
+        emit Swap(msg.sender, nst, gem, wad, out);
+    }
+
+    function swapOut(uint256 wad, uint256 min) external keeper returns (uint256 out) {
+
+        require(block.timestamp >= zzz + hop, "Swapper/too-soon");
+        zzz = block.timestamp;
+
+        require(wad <= maxOut, "Swapper/wad-too-large");
+        require(min >= wad * wantOut / 10**6, "Swapper/min-too-small"); // 1/10**6 = 10**12 / WAD
```

Comment:
Is this written this way (6 decimals) to match USDC?

Reply:
Yes, although I realise that the name `wad` is a bit confusing here as the amount is actually specified in USDC (or USDT) wei's. If we decide to keep this variable (vs using a fixed `lot`), I'd rename `wad` to `amt`.

Reply:
Changed to support other decimal amounts